File: src/lib/sanitize-message-html.ts
'use client';

import DOMPurify from 'dompurify';

/**
 * Санитизация HTML сообщений чата (TipTap) перед выводом через dangerouslySetInnerHTML.
 * Снижает риск XSS при вредоносном содержимом в Firestore.
 *
 * [audit H-010] На сервере (Next.js SSR/RSC) DOMPurify не работает (нет
 * window/DOM), поэтому раньше функция возвращала **сырой** HTML — footgun:
 * любой SSR-кэш мог отправить непроверенное user-supplied содержимое в
 * браузер ДО hydration. Теперь на сервере отдаём пустую строку — UI
 * мигнёт пустотой и сразу гидрируется с очищенным HTML на клиенте.
 * Чат-сообщения требуют auth и не SSR-рендерятся фактически, но
 * гарантия корректнее в коде, чем в комментарии.
 */
export function sanitizeMessageHtml(html: string | undefined | null): string {
  if (html == null || !html.trim()) return '';
  if (typeof window === 'undefined') return '';
  return DOMPurify.sanitize(html, {
    ALLOWED_TAGS: [
      'p',
      'br',
      'strong',
      'b',
      'em',
      'i',
      'u',
      's',
      'strike',
      'del',
      'a',
      'ul',
      'ol',
      'li',
      'blockquote',
      'code',
      'pre',
      'span',
      'div',
      'h1',
      'h2',
      'h3',
      'h4',
    ],
    ALLOWED_ATTR: [
      'href',
      'title',
      'target',
      'rel',
      'class',
      'data-chat-mention',
      'data-user-id',
      'data-chat-custom-emoji',
      'data-emoji-id',
      'data-emoji-src',
      // Animated text effects (Phase 6 mobile/desktop). Web в этой
      // итерации ещё не рендерит анимацию — текст показывается
      // статически, но атрибут сохраняется для round-trip.
      'data-anim',
    ],
    ALLOW_DATA_ATTR: false,
  });
}

File: mobile/app/lib/features/chat/data/sanitize_message_html.dart
import 'package:html/dom.dart' as dom;
import 'package:html/parser.dart' as html_parser;

/// Паритет `src/lib/sanitize-message-html.ts` (DOMPurify allowlist).
const _allowedTags = <String>{
  'p',
  'br',
  'strong',
  'b',
  'em',
  'i',
  'u',
  's',
  'strike',
  'del',
  'a',
  'ul',
  'ol',
  'li',
  'blockquote',
  'code',
  'pre',
  'span',
  'div',
  'h1',
  'h2',
  'h3',
  'h4',
};

bool _attrAllowed(String tag, String attrLower) {
  if (attrLower.startsWith('on') || attrLower == 'style') return false;
  if (attrLower == 'class') return true;
  switch (tag) {
    case 'a':
      return {'href', 'title', 'target', 'rel'}.contains(attrLower);
    case 'span':
      return attrLower == 'data-chat-mention' ||
          attrLower == 'data-user-id' ||
          attrLower == 'data-chat-custom-emoji' ||
          attrLower == 'data-emoji-id' ||
          attrLower == 'data-emoji-src';
    default:
      return false;
  }
}

void _unwrapElement(dom.Element el) {
  final parent = el.parent;
  if (parent == null) {
    el.remove();
    return;
  }
  while (el.nodes.isNotEmpty) {
    parent.insertBefore(el.nodes.first, el);
  }
  el.remove();
}

void _filterAttributes(dom.Element el) {
  final tag = el.localName?.toLowerCase() ?? '';
  final remove = <String>[];
  for (final k in el.attributes.keys.map((k) => k.toString()).toList()) {
    final a = k.toLowerCase();
    if (!_attrAllowed(tag, a)) {
      remove.add(k);
    }
  }
  for (final k in remove) {
    el.attributes.remove(k);
  }
  if (tag == 'a') {
    final href = (el.attributes['href'] ?? '').trim().toLowerCase();
    if (href.startsWith('javascript:') || href.startsWith('data:')) {
      el.attributes.remove('href');
    }
  }
}

void _sanitizeNode(dom.Node parent) {
  for (var i = 0; i < parent.nodes.length;) {
    final child = parent.nodes[i];
    if (child is dom.Comment) {
      child.remove();
      continue;
    }
    if (child is! dom.Element) {
      i++;
      continue;
    }
    final el = child;
    final tag = el.localName?.toLowerCase() ?? '';
    if (tag == 'script' ||
        tag == 'iframe' ||
        tag == 'object' ||
        tag == 'embed') {
      el.remove();
      continue;
    }
    if (!_allowedTags.contains(tag)) {
      _unwrapElement(el);
      continue;
    }
    _filterAttributes(el);
    _sanitizeNode(el);
    i++;
  }
}

/// Санитизация HTML перед отправкой и (опционально) перед рендером.
String sanitizeMessageHtml(String? raw) {
  if (raw == null || raw.trim().isEmpty) return '';
  try {
    final frag = html_parser.parseFragment(raw);
    _sanitizeNode(frag);
    return frag.outerHtml.trim();
  } catch (_) {
    return '';
  }
}
